### ISSUE TITLE
Sound deteriorates on Mac for some audio drivers

### ISSUE THREAD

**Issue by Skeezixes** [jamulussoftware/jamulus#212]
[There may be two issues here. See below.]
If the Settings window is left open for a period of time (15-30 min), and musicians are connected to the server, then there's a good chance that sound quality will go all to wreck (overall delay episodically  > 500ms at the client that is running on the same machine as the server). Once the Overall Delay starts increasing it gets progressively worse and worse. Deteriorated sound quality is also experienced by the other musician (only tried with 1 other) who is connected to the server.

_MacOS, server running on the same machine as the client, wired Ethernet connection._
Replicated. Annotated screenshot is attached.

Possible second issue: Did I have multiple server instances running?? There is no "Server Stopped" message from the first instance, which (I thought) I had killed by closing the Server window. (see the Terminal window in the screenshot). Does closing the Server window not kill that server (because it was launched via the Terminal not the GUI)? If not, shouldn't it?
<img width="1280" alt="Jamulus_0512_annotated" src="https://user-images.githubusercontent.com/64929898/81744460-1e068c80-9458-11ea-9ff0-8c1df7a24361.png">

**Comment by pljones** [comment]
Can you try with the latest Git source for the server issue (not closing).  That should have a fix for it.

**Comment by Skeezixes** [comment]
@pljones Would I have to compile that source? If so, I don't currently have a way to do that (no development environment).

**Comment by pljones** [comment]
OK - that should be fixed when 3.5.3 comes out, at least.

**Comment by Skeezixes** [comment]
More on the Settings window issue:
No matter what else is happening, it appears that when the Settings window is open, the Ping time and the Overall Delay time increase gradually to the point of unusability.

The reason I have the Settings window open, of course, is that I am monitoring Delay (and to a lesser extent, Buffer). How about displaying these times on the client window under the lights? Does anyone else care? Would it be distracting to do this? 

Should it be complicated to solve the underlying issue, is it possible to prevent users from keeping the Settings window open as I have been doing?

**Comment by corrados** [comment]
I leave the Settings window open all the time. I can run Jamulus for hours without any issues. I do this under Windows and Mac. So the question is if this problem only occurs for you or are there any other users affected?

**Comment by corrados** [comment]
BTW, have you tried version 3.5.3? I have just released the new version?

**Comment by Skeezixes** [comment]
Uh oh. Really? What can be the difference between your setup and mine? 
Yes, I am running 3.5.3 now.

**Comment by corrados** [comment]
Do you run the client and server on the same PC? If yes, can you try to shut down the server and just run the client and connect to some remote server. Do you still see the issue then?

**Comment by Skeezixes** [comment]
Yes, they are both on my Mac. I'll try connecting to some other remote server nearby.

**Comment by corrados** [comment]
Ok, thanks for testing. I saw that you already posted in https://github.com/corrados/jamulus/issues/133. Do you think we need two Github issues on this matter or shall we close this and leave the other one open? I would prefer to have just one open Issue about this bug.

**Comment by Skeezixes** [comment]
One issue is better than two. :-) But let me test whether my issue can
happen when the Settings window is closed (#133 doesn’t mention Settings,
but incidentally I am running on the same hardware as he was).

**Comment by corrados** [comment]
BTW: Here is another one which is similar: https://github.com/corrados/jamulus/issues/40

**Comment by corrados** [comment]
One more question:
> Deteriorated sound quality is also experienced by the other musician (only tried with 1 other) who is connected to the server.

Does this mean his own signal is also distorted (which mean the server has an issue) or he just hears your distorted signal (which would mean your client has a problem) but his own audio is still ok?

**Comment by Skeezixes** [comment]
She says that only my sound was bad. Hers was okay.

**Comment by corrados** [comment]
Ok, this is an important information. So it is the client not the server which makes problems in your case. 

> sound quality will go all to wreck

Is it crackeled sound what you here? Can you describe the type of sound quality you get a bit more?

**Comment by Skeezixes** [comment]
It’s sort of hard to describe. I could hear the other player with delays and dropouts and garble that made it extremely difficult but not impossible to stay in time. My sound was usually but not always better so I knew where I was—if you know what I mean—but I didn’t know where I was in relation to the other player. The sound was pretty awful, but I wouldn’t say it was crackled. Does this help?

**Comment by corrados** [comment]
Thanks for the info. I'll try to reproduce it on my Mac. I have not run the server and client in parallel yet. Maybe I can see the issue on my machine as well. That would make the debugging much easier. But I am very busy with other things right now. Do you expect me to do this sometime soon.

**Comment by Skeezixes** [comment]
Of course! Immediately!
I discovered today that there is a server in Portland that is close enough
that we can use it if mine doesn't work right, so whenever you get to it,
right? Sooner is better than later, but we can survive, I think.

**Comment by Skeezixes** [comment]
YES, finally, some insight on what's going on here!
When I have experienced the issue, input = built-in mic.
I just received a new mic. Now, input = USB mic and NO PROBLEM. I just played for > 1 hr on my server, with the Settings window open, and everything was fine. No increased delay, no buffer problem, no sound degradation.

**Comment by corrados** [comment]
Thanks for the additional info. So now we now that it is audio driver related.

**Comment by Skeezixes** [comment]
I played today with one other player, same setup as before (I'm hosting the server, my client running on the same machine, USB mic, wired headphones), and initially the sound was great--super clear, better than usual--but after playing for about 60 minutes the other player's sound deteriorated dramatically over a few seconds with no change in either of our settings, and stayed bad. My sound was fine. 

This bad sound was audible to both the other player and to me, and sounded crackly to both of us, like lots of dropouts, and the other player said her overall delay times were episodically >500 ms. Mine remained within normal range (0-3 ms ping, <21 ms delay). We're both running 3.5.6. My settings window was open; hers was not.

**Comment by corrados** [comment]
That sounds like a temporary network issue. Maybe some PC/Mac/Smartphone did a software upgrade and blocked the network. Then you usually have exactly the situation which you wrote: high ping times and crackly sound.

**Comment by Skeezixes** [comment]
Thanks for the reply. That's good to know.
L

**Comment by pljones** [comment]
Is the issue with "certain audio drivers" on a Mac going to be dealt with in code or documentation?  Is there any other issue here still to address?

**Comment by corrados** [comment]
When I use Jamulus with my Lexicon Omega external USB audio card on my Mac Mini, from time to time, I have the issue that the audio deteriorates and does not recover on its own. Fortunately this does not happen very often. When it happens, I usually quickly change the audio card buffer size to a different value and set it back to the original value. This initiates a reset of the audio driver and the audio is ok afterwards.

I have not found a solution to that problem yet. It seems I do not get continuous xrun messages from Core Audio when it happens. So it is hard to detect this situation from within the Jamulus software. It seems to be some audio driver issue.

**Comment by RonCohen1** [comment]
Hello, I have noticed this same issue, in pretty much every Jamulus session.  And other people in these sessions observe the same thing -- audio degrades; change buffer size (or even just click on the same buffer size), things get better for a while, then deteriorate again.  This seems to not happen on another platform (Soundjack), so it seems specific to Jamulus, and possibly specific to the Mac version (I will make further queries to figure that out).  Maybe there is a memory leak?

**Comment by gene96817** [comment]
I think I am also having a similar problem. My hypothesis is there is an accumulating delay in one or more of the channels. Is there a way to force all the channels to resync (reset)? If there is a way to stop and restart all the channels without disconnecting, then we can see if the defect is accumulating skew.

**Comment by RonCohen1** [comment]
I have some more data.  I took a survey of singers in two groups that are using Jamulus regularly for rehearsal.  It seems only two of us, both on Macs, are experiencing this problem.  I've experienced it on both my 11-year-old Macbook Pro and the new one I got in late spring.  But the surveyed groups include Mac users with no problems.  So it must be associated with some particular system setting.  I'm hoping someone can suggest something to try.

**Comment by gene96817** [comment]
There are so few system settings affecting the audio. How would you distinguish between your system setting and an internet condition? I wouldn't assume the internet connection is not a possible suspect.

**Comment by RonCohen1** [comment]
One thing that comes to mind is background system processes -- e.g. anti-virus software, or anything else not a standard part of the OS.  I am running Symmantec anti-virus.  Also iStat menus (provides continuous info on system load etc.) I am wondering if you run either of those (or another anti-virus program).  RE internet connection: I have quite fast internet.  Internet problems could for sure cause dropouts, etc, but it's hard to see how that would be a problem that would always increase during a Jamulus session.

**Comment by gene96817** [comment]
I don't run any anti-virus. Instead, I run Block-Block, a host-based intrusion protection system (H-IPS). I do have to watch out for all my programs that run sync (i.e. dropbox, Google Drive, backup, and iCloud). I try to remember to turn them off. I should have plenty of CPU capacity for Jamulus.

I am working with my ISP to resolve a very slow network (95Mbps instead of 1Gpbs). But 95 Mbps is supposed to be fast enough. Besides speed, I suspect a bottleneck related to bufferbloat. I have a (hypothetical) model that TCP congestion affects Jamulus UDP and can cause accumulation of delay on some channels. None of the developers believe this model. I am trying to eliminate this as a suspect.

If you eliminate the possibility of a rogue process on your Mac, then my Internet suspect might be affecting you.

**Comment by RonCohen1** [comment]
I think I've read elsewhere the suggestion of bufferbloat.  (Or maybe that was you?).  And maybe in connection with that, that the problem lies in the modem or router.  Though I've also changed the modem/router since starting on Jamulus, again with no change.   By the way your 95Mbps should be fine.   I upgraded my internet connection from 90 Mbps to 650, with no change in Jamulus behavior (no surprise).

**Comment by gene96817** [comment]
It is not the throughput as measured by speed. What I have measured is variable ping delays... on good days <10 ms.   On "common" days, 30-40 ms. On bad days, 80-150 ms. All of these different delays can show the same throughput. Typical TCP applications (email, Netflix, etc.) don't care. Typical UDP applications care.

**Comment by RonCohen1** [comment]
Highly variable ping time will certainly contribute to poor sound quality as packets will arrive seriously out of order, so without very large jitter buffer settings there's no way for Jamulus to re-arrange the sequence.  So, garbled sound.  But even then, why should the problem get worse with time?    What kind of internet do you have? Your highly variable ping times seem characteristic of DSL connections.   I have cable internet; The ping time to my private (cloud) server is no where nearly as variable.  Usually jumping around between 11-16 ms; on a good day it might get as low as 6 ms, and on a bad day as high as 22.  But still I have this problem of the noise getting worse with time.

**Comment by gene96817** [comment]
Does Jamulus have a mechanism for reordering packets? It is not part of the UDP frame. Jamulus would have to enumerate the UDP payload. If the packets are enumerated, we can add a little logic to know if a packet is so late that it should be discarded. Being able to do this discard would enable flushing a channel of late packets to get it back in sync with the other channels.

I (am supposed to) have 1 Gbps fiber service. Getting only 95 Mbps. That plus the highly variable ping times is a sure sign of a bottleneck or an overloaded router. I will get this service problem fixed.

I believe the highly variable ping times and overloaded router has exposed a perverse condition when the router buffers are blocked. The UDP packets are delivered late. If there is no mechanism for dropping late packets, then Jamulus will process all the UDP packets, regardless of how late. Then that channel will be permanently late, getting later and later with each buffer blockage. This is just a hypothesis. There is no data to prove or disprove this is occurring.

If the UDP payload is already enumerated, it would be possible to add some logic to confirm this condition.

**Comment by RonCohen1** [comment]
My understanding is Jamulus implemented something to re-order out-of-order packets as of Version 3.6.0.  I like your theory about the late packets; that would explain the phenomenon, though I have no way of knowing if the condition you posit is actually the case.  It would be nice if Werner or another developer could offer a comment.

As it happens, I connected to a session this morning as a listener, and while on the session, turned off my anti-virus.  It seemed to help.  I've now also turned off iStat menus.  I have a rehearsal later today; it will be interesting to see if these steps make a difference.

I don't know if I've mentioned it in this thread, but I seem to experience better behavior on another platform (Soundjack).  But since that is primarily peer-to-peer (their server capability is in a fairly primitive state of development), it's hard to make a definitive statement.

**Comment by gene96817** [comment]
I would be glad to discuss strategies for using the enumeration to understand the audio defects. 

You can use Activity Monitor to determine if you have enough spare CPU cycles to leave some of those other task running.

It is easy for a new early-stage code to be fast and nimble. As the code becomes more fully-featured, those features will consume CPU cycles. A motorcycle can be fast and nimble but it is not a car.

**Comment by RonCohen1** [comment]
Right, I know server-based models scale better to larger number of singers.  But what is true is for just a few singers, where point-to-point makes sense, Soundjack works great -- about half the latency (as you're connected directly to the other peers, rather than via a server); and I don't get the annoying accumulation of sound garbage.   RE CPU cycles, for Jamulus I never get above about 8% of CPU devoted to Jamulus on my Mac.  Since I shut down everything I can, the CPU is mostly idle when I have Jamulus running.  

I was on another rehearsal tonight, with anti-virus and iStat menus disabled.   Still the accumulation of noise/crackles etc.   Since I have a half-year-old quite capable MacBook Pro, and since others running older, less capable machines seems not to have this problem I have to assume the problem is not my computer.  So I wonder if the problem can be in the router or modem -- though I am using an integrated gateway that is also new, and capable of GB+ speeds.   So where else can I look?

**Comment by gene96817** [comment]
I have been thinking about the accumulated sound problem as a server issue.

Your MBP shouldn't be the problem. Running at <10% load shouldn't cause any delays. 
An integrated router/modem is usually pretty recent. I would check for any firmware updates. If it has an update (bug fix) it will make a difference. If you have a lot of port forwarding defined, that could make a difference.

To me, noise/crackles are caused by dropped, late, or out of order packets. (Since you report Jamulus reorders packets, out-of-order packets is not the problem.) With your MBP running <10% loading, I would look at your transport. Currently, the only transport-related clues you would have is the fluctuating latency and total delay numbers. Use the largest buffer setting you can. If the latency fluctuation is larger than the buffer time equivalent, every time you get a big latency bounce, you are likely to get a buffer underrun followed by a buffer overrun (which is a lost packet).

**Comment by RonCohen1** [comment]
Two thoughts: (1) I didn't notice a big change when Jamulus 3.6.0 came out, so it's not clear to me that Jamulus's ordering of packets was completely effective.  The release notes didn't provide any detail.   (2) I looked at my gateway's firewall log just a short while ago and found over 11000 WANATTACK drop attempts this evening.  Along with 3629 PPv6 FORWARD drops and 2736 IPv6 input drop attempts.   I have no idea what generates these, but it seems like something that could tie up the gateway and interfere with orderly packet transmission.   My observation though is that when this packet loss phenomenon happens, it only affects what I hear, not what others hear of me.   RE firmware updates, this is the cable company's router, I have no means to check for, let alone install, a firmware update.

**Comment by gene96817** [comment]
1- I am curious about what the packet ordering algorithm does. My wild guess is it is of limited effect.
2- You are getting bursts of scans that are being blocked by your router/firewall.  While you are getting a burst of scans, it would affect your inbound Jamulus traffic and that is the disruption you are hearing. Your outbound traffic is not affected (until you get a big burst of inbound representing a DOS attack). Nothing you can do about these. At least they are being blocked by your firewall.
3- your router has a web management interface (the one you used to get those firewall logs).  In the management/maintenance section will be a place for triggering a firmware update. That is the only update I would recommend. It's pretty idiot-proof.

**Comment by RonCohen1** [comment]
Re your comments: 1.  I agree.  2. The bursts of scans can account for inbound packets not making it to my computer in a timely fashion, which would certainly impact audio.  Unclear to me why this effect should increase in time, and why resetting the Jamulus buffer delay would make it temporarily better.   But anyway I am going to contact my ISP to see if there is anything that can be done to alleviate the attacks.  3.  There is no place in the gateway's management interface to trigger a firmware update.  Apparently the  gateway provider (Comcast) rolls out firmware updates automatically.

**Comment by gilgongo** [comment]
Hi all - I'm moving this to a discussion until we can identify the thing(s) that need doing to fix this, at which point we can raise the appropriate work tickets.
